### ISSUE TITLE
Reordering is breaking statements with side-effects.

### ISSUE THREAD

**Issue by mbostock** [rollup/rollup#34]
Here’s linear.js (abridged):

```js
export function Linear() {};
Linear.prototype.lineStart = function() {};
Linear.prototype.lineEnd = function() {};
Linear.prototype.point = function(x, y) {};

export function LinearClosed() {};
LinearClosed.prototype = Object.create(Linear.prototype);
LinearClosed.prototype.lineEnd = function() {};
```

The resulting rolled-up file:

```js
function Linear() {}
Linear.prototype.lineStart = function() {};
Linear.prototype.lineEnd = function() {};
Linear.prototype.point = function(x, y) {};

function LinearClosed() {}
LinearClosed.prototype.lineEnd = function() {};
LinearClosed.prototype = Object.create(Linear.prototype);
```

Note that the last two lines have been reversed, so the assignment to LinearClosed.prototype.lineEnd gets overwritten. This code is in [d3-shape](https://github.com/d3/d3-shape). I can work on creating an isolated test case if you need it.

**Comment by Rich-Harris** [comment]
Thanks. No need for a more isolated test, I reckon this is small enough to work with - will look into it as soon as I get a chance. Not immediately sure what's causing it but am sure it's fixable (he said confidently...)

**Comment by Rich-Harris** [comment]
@mbostock Turns out I'm struggling to reproduce this - all the d3-shape tests are passing for me, and the statement ordering looks to be correct. Has anything changed since this issue was opened?

**Comment by mbostock** [comment]
Yeah, there’s [a workaround](https://github.com/d3/d3-shape/blob/master/src/interpolate/linear.js#L17) I’ve been using.

I’ll try to pare down a standalone test case.

**Comment by mbostock** [comment]
Here ya go:

https://gist.github.com/mbostock/bc1ffc78c3f933eff312

Note that if you reverse the order of statements in index.js, the test passes!

**Comment by Rich-Harris** [comment]
Ah, I see what's going on here. The `(new Foo)` line causes the `Foo` definition to be included, which means also including `Foo.prototype.test = ...` and (because rollup can't know that `Object.create` doesn't modify `Foo.prototype`) `Bar.prototype = Object.create(Foo.prototype)`. So the second line, which causes the `Bar` declaration to be included, *doesn't* include the `Bar.prototype =` statement because it's already included.

Not immediately sure what a general solution looks like (the last time I tried to naively sort statements according to their original order, it [ended badly](https://github.com/rollup/rollup/issues/21)), but I'll try a few things. Suggestions welcome!

**Comment by Rich-Harris** [comment]
Fixed in 0.9.0 - have confirmed all the d3-shape tests now pass without the workaround
